lib/logdna/client.rb

```diff
@@ -1,164 +1,134 @@
+        processed_message = process_message(msg, opts)
+        new_message_size = processed_message.to_s.bytesize
+        @buffer.push(processed_message)
+        @buffer_byte_size += new_message_size
+        @flush_scheduled = true
+        @lock.unlock
+
+        flush if @flush_limit <= @buffer_byte_size
+        schedule_flush unless @flush_scheduled
       else
-        @task = create_flush_task
+        @side_message_lock.synchronize do
+          @side_messages.push(process_message(msg, opts))
+        end
       end
     end
 
-    # this should be running synchronously if @buffer_over_limit i.e. called from self.buffer
-    # else asynchronously through @task
-    def flush()
-      if defined? @@request and !@@request.nil?
-        request_messages = []
-        @lock.synchronize do
-          request_messages = @messages
-          @buffer.truncate(0)
-          @messages = []
-        end
-        return if request_messages.empty?
+    def send_request
+      @side_message_lock.synchronize do
+        @buffer.concat(@side_messages)
+        @side_messages.clear
+      end
 
-        real = {
-          e: 'ls',
-          ls: request_messages,
-        }.to_json
+      @request.body = {
+        e: "ls",
+        ls: @buffer
+      }.to_json
 
-        @@request.body = real
-        @response = Net::HTTP.start(@uri.hostname, @uri.port, use_ssl: @uri.scheme == 'https') do |http|
-          http.request(@@request)
+      handle_excpetion = lambda do |message|
+        puts message
+        @exception_flag = true
+        @side_message_lock.synchronize do
+          @side_messages.concat(@buffer)
         end
+      end
 
-        # don't kill @task if this was executed from self.buffer
-        # don't kill @task if there are queued messages
-        unless @buffer_over_limit || @messages.any? || @task.nil?
-          @task.shutdown
-          @task.kill
+      begin
+        @response = Net::HTTP.start(
+          @uri.hostname,
+          @uri.port,
+          use_ssl: @uri.scheme == "https"
+        ) do |http|
+          http.request(@request)
+        end
+        if @response.is_a?(Net::HTTPForbidden)
+          puts "Please provide a valid ingestion key"
+        elsif !@response.is_a?(Net::HTTPSuccess)
+          puts "The response is not successful "
         end
+        @exception_flag = false
+      rescue SocketError
+        handle_excpetion.call("Network connectivity issue")
+      rescue Errno::ECONNREFUSED => e
+        handle_excpetion.call("The server is down. #{e.message}")
+      rescue Timeout::Error => e
+        handle_excpetion.call("Timeout error occurred. #{e.message}")
+      rescue
+        handle_excpetion.call("#{e.message}")
+      ensure
+        @buffer.clear
       end
     end
 
-    def exitout()
-      check_side_buffer
-      if @messages.any?
-        flush()
+    def flush
+      if @lock.try_lock
+        @flush_scheduled = false
+        return if @buffer.empty? && @side_messages.empty?
```

Comment:
This seems wrong. You're returning without releasing the `@lock`. I think this could happen if 2 threads schedule a flush, one gets there first so empties the buffer/side_messages, then this thread gets here and never unlocks.

Reply:
good point. Thanks